### ISSUE TITLE
Implement a "desugaring" pass that turns Fe code into other Fe code

### ISSUE THREAD

**Issue by cburgdorf** [ethereum/fe#205]
### Abstract

Compilers often implement a desugaring pass where certain features are just syntactic sugar for a more verbose form.

For instance, if we have support for structs than tuples can be implemented as syntactic sugar that internally gets converted to a struct.

For instance, a Tuple of type `(u8, string100)` may just be syntactic sugar such a struct:

```
struct $tuple_u8_string100:
    0: u8
    1: string100
```

Another example would be rewriting string literals to constants such as:

```
def foo(val: string100):
    if val == "foo":
        pass
```

```
def foo(val: string100):
    if val == "$CONST_STRING_FOO:
        pass
```

**Comment by cburgdorf** [comment]
@g-r-a-n-t I just did a quick and dirty write-up because I wanted to fill our February milestone with issues and know that you'd start working on this soon.

**Comment by g-r-a-n-t** [comment]
Thanks for the writeup. For reference, I'm linking the closure [desugaring post](https://rustyyato.github.io/rust/syntactic/sugar/2019/01/17/Closures-Magic-Functions.html) here.

**Comment by g-r-a-n-t** [comment]
closed in #338
